### PR TITLE
Add and implement Player#openBook API

### PR DESCRIPTION
**Pull request by jacobsandersen**
This PR adds an API to open books in `org.bukkit.entity.Player`. The `Player#openBook` method accepts an `org.bukkit.inventory.ItemStack` and is implemented in `org.bukkit.craftbukkit.CraftPlayer#openBook`. The method expects a non-null bukkit `ItemStack` with the material type `Material.WRITTEN_BOOK`.

`net.minecraft.server.EntityPlayer` was modified to include an OBFHELPER, forwarding the existing method `EntityPlayer#a(ItemStack itemstack, EnumHand enumhand)` to `EntityPlayer#openBook(...)`.

The implementation operates on the requirement for the book to be opened to be placed in the player's main hand (off hand ALSO works, but I did not include it in this PR; but can add.) Firstly, the item in the main hand is saved. Then, the requested book is placed into the Player's main hand. Thirdly, `EntityPlayer#openBook` is called with the book converted to an NMS copy. Finally, the player's original main hand is restored.

An alternative to passing in an ItemStack: the implementation could be modified to read the player's main hand and have a precondition on the `Material`.

Example:
```java
@EventHandler
public void onClick(InventoryClickEvent event) {
  Player player = (Player) event.whoClicked;
  ItemStack book = new ItemStack(Material.WRITTEN_BOOK);
  
  BookMeta meta = (BookMeta) book.getItemMeta();
  meta.spigot().addPage(
    new ComponentBuilder("You clicked and opened a book!")
      .color(ChatColor.RED)
      .bold(true)
      .event(
        ClickEvent.Action.OPEN_URL,
        "https://google.com"
      )).create()
  );

  book.setItemMeta(meta);
  player.openBook(book);
}
```
Preview:
https://i.imgur.com/RWI6DSN.gifv

### CODE REVIEW

**Comment by Black-Hole** [comment]
You could fit the OPFHELPER in one line and move the @Override in front of the line after next.

File: Spigot-Server-Patches/0400-Implement-open-book-api.patch [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+     @Override
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+index 9f267414d..82dcb6829 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+@@ -66,6 +66,7 @@ import net.minecraft.server.TileEntitySign;
+ import net.minecraft.server.Vec3D;
+ import net.minecraft.server.WhiteListEntry;
+ import net.minecraft.server.WorldServer;
++
```
Comment (Spottedleaf) [inline Spigot-Server-Patches/0400-Implement-open-book-api.patch]: newline

**Comment by jacobsandersen** [comment]
> You could fit the OPFHELPER in one line and move the @override in front of the line after next.

Thanks.

File: Spigot-Server-Patches/0400-Implement-open-book-api.patch (hunk 1 above)
Comment (jacobsandersen) [reply Spigot-Server-Patches/0400-Implement-open-book-api.patch]: done

**Comment by jacobsandersen** [comment]
I've migrated this PR to upstream at https://hub.spigotmc.org/stash/projects/SPIGOT/repos/bukkit/pull-requests/444/overview and https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/pull-requests/571/overview

**Pull request closed without merging**